### PR TITLE
Add support for additional user attributes to TomcatPrincipal

### PR DESCRIPTION
**Pull request by cklein05**
Add support for additional user attributes to the `TomcatPrincipal` interface and the `GenericPrincipal` class.

### CODE REVIEW

File: java/org/apache/catalina/TomcatPrincipal.java [hunk 1]
```diff
@@ -47,4 +48,37 @@
...
      */
     void logout() throws Exception;
+
+    /**
+     * Returns the value of the named attribute as an <code>Object</code>, or
+     * <code>null</code> if no attribute of the given name exists, or if
+     * <code>null</code> has been specified as the attribute's name.
+     * <p>
+     * Only the servlet container may set attributes to make available custom
+     * information about a Principal or the user it represents. For example, some of
```
Comment (michael-o) [inline java/org/apache/catalina/TomcatPrincipal.java:58]: I think the For example is not ncessary, since we don't want to incline to any actual implementation.

File: java/org/apache/catalina/TomcatPrincipal.java [hunk 2]
```diff
@@ -47,4 +48,37 @@
...
+     * the Realm implementations can be configured to additionally query user
+     * attributes from the <i>user database</i>, which then are provided through the
+     * Principal's attributes map.
+     * <p>
+     * Attribute names and naming conventions are maintained by the Tomcat
+     * components that contribute to this map, like some of the Realm
+     * implementations.
```
Comment (michael-o) [inline java/org/apache/catalina/TomcatPrincipal.java:65]: Are those really maintained? I thought the dev/admin requests the realm to load attribute values. So the attribute names are not necessarily mandated?!

File: java/org/apache/catalina/realm/GenericPrincipal.java [hunk 3]
```diff
@@ -171,6 +176,7 @@ public GenericPrincipal(String name, List<String> roles,
         }
         this.loginContext = loginContext;
         this.gssCredential = gssCredential;
+        this.attributes = attributes;
```
Comment (michael-o) [inline java/org/apache/catalina/realm/GenericPrincipal.java:179]: Should we wrap with an unmodifiable map or do we expect the caller to do this?

File: java/org/apache/catalina/realm/GenericPrincipal.java [hunk 4]
```diff
@@ -283,10 +294,16 @@ public boolean hasRole(String role) {
...
         sb.append(this.name);
         sb.append('(');
         for (String role : roles) {
-            sb.append(role).append(',');
+            if (first) {
+                first = false;
+            } else {
+                sb.append(',');
+            }
+            sb.append(role);
```
Comment (michael-o) [inline java/org/apache/catalina/realm/GenericPrincipal.java:306]: While this is correct, it should be in separate PR.

**Review by michael-o — CHANGES_REQUESTED** [review]
The first and foremost question I have: Is this save to be backported to previous Tomcat versions?

File: java/org/apache/catalina/realm/GenericPrincipal.java (hunk 3 above)
Comment (cklein05) [reply java/org/apache/catalina/realm/GenericPrincipal.java:179]: AFAIK we agreed on not using an unmodifiable map, since there is no write access to this map. However, we could, of course, if you like so. Maybe it's worth checking if it's already an instance of `java.util.Collections.UnmodifiableMap` first. If not, we could wrap it.

File: java/org/apache/catalina/TomcatPrincipal.java (hunk 2 above)
Comment (cklein05) [reply java/org/apache/catalina/TomcatPrincipal.java:65]: Don't know whether it's really the exact word. However, every attribute provider (aka Realm) can use it's own naming schema and conventions. Some may use prefixes (user.displayName) others may not, who knows. We make no rules or assumptions here but only refer to the each attribute provider's documentation.

File: java/org/apache/catalina/TomcatPrincipal.java (hunk 1 above)
Comment (cklein05) [reply java/org/apache/catalina/TomcatPrincipal.java:58]: Will remove that.

File: java/org/apache/catalina/realm/GenericPrincipal.java (hunk 3 above)
Comment (cklein05) [reply java/org/apache/catalina/realm/GenericPrincipal.java:179]: Class `java.util.Collections.UnmodifiableMap` is private, so we cannot check instance.

**Comment by cklein05** [comment]
> Is this save to be backported to previous Tomcat versions?

It should be. I do not see any reasons why this cannot be backported.

File: java/org/apache/catalina/TomcatPrincipal.java (hunk 2 above)
Comment (cklein05) [reply java/org/apache/catalina/TomcatPrincipal.java:65]: In other words, when querying an attribute named "displayName", a Realm could add this under a key like "realm.displayName" or "user.displayName". For that, the (maybe custom) Realm may have a config option `userAttributePrefix` or this could even be hard-coded.

Such prefixes might make sense, if, in the future, for example, several components could contribute to the Principal's attributes (like the Authenticator or a new component `SocialUserDataProvider` querying user information from Facebook etc.).

**Comment by michael-o** [comment]
> > Is this save to be backported to previous Tomcat versions?
> 
> It should be. I do not see any reasons why this cannot be backported.

The big question is here is whether the `TomcatPrincipal` is truly limited to our Realms and not intended to be used in custom Realms. Because if custom Realm will use this interface a mere Tomcat will break their implementations. @markt-asf Can we make this assumption that this is Tomcat private?

**Comment by cklein05** [comment]
We could always remove the new _attributes_ methods from the `TomcatPricipal` interface and work with class `GenericPrincipal` only. In other words, having these methods declared on the interface is not crucial for the enhancement. In that case, we just need to update the `instanceof` check in the modified JSP example at /examples/jsp/security/protected as well as the documentation (and example code) under _Common Features / Additional User Attributes_ in realm-howto.xml (both not yet committed).

BTW, shouldn't we commit/merge the example mentioned above with this PR? Listing the Principal's additional attributes is completely independent from the Realm implementations. The same is likely true for the new _Common Features / Additional User Attributes_ section in realm-howto.xml.

**Comment by markt-asf** [comment]
If we limit the back-port to Tomcat 9+, we are on Java 8 and can use default methods to maintain compatibility.

**Comment by cklein05** [comment]
> If we limit the back-port to Tomcat 9+, we are on Java 8 and can use default methods to maintain compatibility.

I guess, that's the way to go. Shall I implement these in my branch already? However, I will commit all changes only after we've agreed on all open questions.

**Comment by michael-o** [comment]
@cklein05 Please apply all requested changes. I want to merge with main and proceed with Tomcat 9+ as @markt-asf has written.

**Comment by michael-o** [comment]
> What about the example JSP application? Shouldn't we add support for dumping user attributes with this commit? See [#463 (comment)](https://github.com/apache/tomcat/pull/463#issuecomment-1015511489) (2nd paragraph)

Personally, I wouldn't do this. This will limit the semantics of your PR, taking away to work with interface only. My custom principal uses this one as well so Tomcat can internally use the provided methods otherwise it feels foreign to Tomcat.

**Comment by cklein05** [comment]
I'm not talking about removing methods from the interface. In the 2nd paragraph of this comment I was asking whether to include the additions to Tomcat's example JSP application located under `/examples/jsp/security/protected`. The modified example application supports showing any additional user attributes of the authenticated Principal.

**Comment by michael-o** [comment]
> I'm not talking about removing methods from the interface. In the 2nd paragraph of this comment I was asking whether to include the additions to Tomcat's example JSP application located under `/examples/jsp/security/protected`. The modified example application supports showing any additional user attributes of the authenticated Principal.

My bad, yes. Please reuse the example as well.

**Comment by cklein05** [comment]
Done.

**Comment by michael-o** [comment]
@cklein05 The default methods are on not required on main since it is still bleeding edge, we can break API here. Default methods are required for backports to 9 and 10 only.

**Comment by cklein05** [comment]
Removed

**Comment by michael-o** [comment]
> @michael-o Will provide a fix for _Travis CI failed_ ASAP.

Thanks, will then run it locally.

**Comment by cklein05** [comment]
@michael-o Should be fine now.

**Comment by michael-o** [comment]
@cklein05 Please provide backport PRs for 10.0.x and 9.0.x with default methods.
